test/x86/avx512f.c

```diff
@@ -1309,6 +1534,214 @@ test_simde_mm512_setone_si512(const MunitParameter params[], void* data) {
+                              UINT64_C(0x9f43b81b73f10ecc), UINT64_C(0x5cf361258c69074e),
+                              UINT64_C(0xff40b7e681a97da7), UINT64_C(0xa0db6e89c3bba587),
+                              UINT64_C(0x1ef0d82a3032940e), UINT64_C(0x6e6129fbad8436eb)),
+      UINT32_C(26),
+      simde_x_mm512_set_epu64(UINT64_C(0x0000000ef9c9533d), UINT64_C(0x0000000f8c191bbb),
+                              UINT64_C(0x00000027d0ee06dc), UINT64_C(0x000000173cd84963),
+                              UINT64_C(0x0000003fd02df9a0), UINT64_C(0x0000002836dba270),
+                              UINT64_C(0x00000007bc360a8c), UINT64_C(0x0000001b984a7eeb)) },
+    { simde_x_mm512_set_epu64(UINT64_C(0x8c60b9ee3084e68a), UINT64_C(0x7df27a77163f440e),
+                              UINT64_C(0x1c8e2ce2cbcdb26b), UINT64_C(0x9a875b341e54f334),
+                              UINT64_C(0xadd74f036b2fd29e), UINT64_C(0x4df18e156438f1fb),
+                              UINT64_C(0x006d02efd4f52b83), UINT64_C(0xc04924106403aeb8)),
+      UINT32_C(27),
+      simde_x_mm512_set_epu64(UINT64_C(0x000000118c173dc6), UINT64_C(0x0000000fbe4f4ee2),
+                              UINT64_C(0x0000000391c59c59), UINT64_C(0x0000001350eb6683),
+                              UINT64_C(0x00000015bae9e06d), UINT64_C(0x00000009be31c2ac),
+                              UINT64_C(0x000000000da05dfa), UINT64_C(0x000000180924820c)) },
+    { simde_x_mm512_set_epu64(UINT64_C(0xb37b4d212858a1f2), UINT64_C(0x332c5a15c349f1f8),
+                              UINT64_C(0x89c2363c78a22d75), UINT64_C(0x7a88d63e31490779),
+                              UINT64_C(0x49a6fa27e84df7f6), UINT64_C(0x7a2a8b04c05b79dc),
+                              UINT64_C(0x2cb04d0b912c9a91), UINT64_C(0x38ff445c61499eb7)),
+      UINT32_C(5),
+      simde_x_mm512_set_epu64(UINT64_C(0x059bda690942c50f), UINT64_C(0x019962d0ae1a4f8f),
+                              UINT64_C(0x044e11b1e3c5116b), UINT64_C(0x03d446b1f18a483b),
+                              UINT64_C(0x024d37d13f426fbf), UINT64_C(0x03d154582602dbce),
+                              UINT64_C(0x016582685c8964d4), UINT64_C(0x01c7fa22e30a4cf5)) },
+    { simde_x_mm512_set_epu64(UINT64_C(0xef4891f108bd3572), UINT64_C(0x70ed11eef2e11aa0),
+                              UINT64_C(0xef3984671f62c16d), UINT64_C(0x00cdcea30b923238),
+                              UINT64_C(0x44ef5f23e94bd07e), UINT64_C(0x61a977d25ae9865b),
+                              UINT64_C(0xbe912feb87b529cc), UINT64_C(0x6a55790501f02cd4)),
+      UINT32_C(1),
+      simde_x_mm512_set_epu64(UINT64_C(0x77a448f8845e9ab9), UINT64_C(0x387688f779708d50),
+                              UINT64_C(0x779cc2338fb160b6), UINT64_C(0x0066e75185c9191c),
+                              UINT64_C(0x2277af91f4a5e83f), UINT64_C(0x30d4bbe92d74c32d),
+                              UINT64_C(0x5f4897f5c3da94e6), UINT64_C(0x352abc8280f8166a)) },
+  };
```

Comment:
It's probably a good idea to include at least one value outside of [0, 255] for `b`.  For example, 1066, which should shift by 42 bits (`1066 & 0xff == 42`).

Reply:
You're good at logical operations on decimals ;-)
The example I based this on only used a fixed shift value of 7 (IIRC) :-(
(Maybe this was because some functions like this need the shift value to be a compile-time constant but I don't think this was the case here.)
I decided that this wasn't sufficient coverage.
I deliberately forced 64 because sometimes shift operators take the shift count modulo the number of bits.
I'll go further if you wish.

Reply:
Not really, but my calculator is.  I'm too lazy to do it by hand :)

Yeah, 64 (or really anything in [64, 255]) is good to have, but you're still not really checking to make sure the upper 3 bytes are ignored.  Actually, you might want to do one where `b & 0x3f == b & 0xff` (like 1066; maybe generate it with `munit_rand_memory(sizeof(b), &b); b &= 0xc0;`), and one where `b & 0x3f != b & 0xff`?  That way you also cover the case of having to ignore high bits *and* make sure the shift isn't modulo n bits.

Reply:
> It's probably a good idea to include at least one value outside of [0, 255] for `b`. For example, 1066, which should shift by 42 bits (`1066 & 0xff == 42`).

Houston we have a problem :-
As you say, according to the IIG, your 1066 (0x42a) should be equivalent to 42 (0x2a).
In fact shifting by 1066 always gives a zero result :-(
So it looks as if more than 8 bits are taken into account.
(Values > 16 bits are also taken into account.)

This is in /x86/avx512f/mm512_srli_epi64/native/c when running under the Intel SDE.
Or am I being stupid again?

Reply:
I attach quick and dirty test code to prove the IIG is wrong.
It is very slow under the Intel SDE so I only go up to 24 bits.
You can just change MAX if you are lucky enough to have real AVX512 hardware.
I thought I could attach it but it complained.
The formatting suffers a bit in "Preview", sorry.
So here it is inline:-

#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <immintrin.h>

#if 0
#define MAX ((unsigned long) UINT_MAX + 1)
#else
#define MAX (1UL << 24)
#endif

int main(void)
{
  __m512 ff;
  unsigned long i;
  unsigned int index, last_index = 0;
  int last = 0, is_zero, last_is_zero = 0;

  ff = _mm512_set1_epi8(0xff);

  for (i = 0; !last; ++i)
    {
      last = (i >= MAX);
      index = (unsigned int) i;

      is_zero = ! _mm512_test_epi64_mask(_mm512_srli_epi64(ff, index), ff);

      if ((is_zero != last_is_zero) || last)
	{
	  printf("%s from %u to %u\n", last_is_zero ? "zero" : "non zero", last_index, index - 1);
	  last_is_zero = is_zero;
	  last_index = index;
	}
    }

  return EXIT_SUCCESS;
}

Reply:
It looks like the Intrinsics Guide is wrong.  I just used this:

```c
#include <immintrin.h>
#include <stdint.h>
#include <inttypes.h>
#include <stdlib.h>
#include <stdio.h>

int main(void) {
  __m512i v = _mm512_set1_epi64(~INT64_C(0));
  __m512i r1 = _mm512_srli_epi64(v, 0x042a);
  __m512i r2 = _mm512_srli_epi64(v, 0x042a & 0xff);

  printf("0x%016" PRIx64 "\n", *((int64_t*) (&r1)));
  printf("0x%016" PRIx64 "\n", *((int64_t*) (&r2)));

  return EXIT_SUCCESS;
}
```

To [test on AVX-512 hardware](https://travis-ci.org/github/nemequ/ci-noise/builds/671320307) (Travis CI often runs on hardware which supports AVX-512).

The good news is you don't need to worry about testing those values, or masking the input :)

Do you want me to take care of notifying Intel, or would you like to do it?

Reply:
Yes, I imagine that you have more contacts at Intel than I do.
The problem is worse with the 256 and 128 versions of these instructions.
Not only do they have this problem but they also take an int as the shift count and the pseudo-code doesn't test for negative values. They should change the shift count to unsigned int.
There are probably heaps of other intrinsics with similar problems and I just stumbled on this one.

Reply:
> I made a few changes prior to merging. It's all trivial stuff, normally I'd just silently "fix" it and thank you for the patch but since it seems like you're interested in contributing going forward I thought it would be a good idea to let you know what I changed and why.
> 
>     * For the preprocessor stuff, I changed the indentation to basically treat it just like the C code.  I know there isn't any consistency on this in SIMDe, but I've been trying to slowly move to this style; it makes it easier to generate code programmatically which is formatted correctly.
> 

I don't really agree with these preprocessor modifications.
Unless this has changed in the last 40 years, strictly the # of the preprocessor directive **MUST** be in column 1 although few preprocessors enforce this. However the # can be separated from the keyword.

```
// These are OK
#         define TEST 1
#         if TEST

// but these are not
         #define TEST 1
         #if TEST
```

>     * In the xor function, I switched the portable fallback to use the i32f member (`int_fast32_t`), which may be slightly faster than `int64_t`.
>

OK

>     * Removed the MODIF_CM checks for the fallbacks.  I'm guessing this is an artifact from some other changes.
>

So sorry about this.
I thought I had removed them all.
I obviously didn't look at the final diff carefully enough.
They were for me to easily test the speed difference with and without the fallbacks.

>     * Used `*mem_addr` instead of `simde_memcpy` in `simde_mm512_load_si512`.  This function is for aligned loads, so it's okay.  The `simde_assert_aligned` call theoretically makes this unnecessary since it uses the [`HEDLEY_ASSUME`](https://github.com/nemequ/hedley/blob/d9261eeb6907b502dd06581932a6738ce4f65012/hedley.h#L1113) to tell the compiler that the data is aligned, but not all compilers support it.
>

OK

>     * Switched from C++ style comments to C style comments (`/* ... */` instead of `// ...`).  Again, this shouldn't really matter since we already require C99 (or at least the subset MSVC supports) and C++ style comments are included in C99, but it does keep our C99 dependency weaker.
>

Noted

Reply:
> Unless this has changed in the last 40 years, strictly the # of the preprocessor directive MUST be in column 1 although few preprocessors enforce this. However the # can be separated from the keyword.

C99 and up pretty clearly allow it.  [§ 6.10 ¶ 2 of the C99 spec](http://port70.net/~nsz/c/c99/n1256.html#6.10p2) (emphasis added):

> A preprocessing directive consists of a sequence of preprocessing tokens that satisfies the following constraints: **The first token in the sequence is a # preprocessing token that** (at the start of translation phase 4) is either the first character in the source file (optionally after white space containing no new-line characters) **or that follows white space containing at least one new-line character**. The last token in the sequence is the first new-line character that follows the first token in the sequence. A new-line character ends the preprocessing directive even if it occurs within what would otherwise be an invocation of a function-like macro.

AFAIK C89 was ambiguous about this (and pretty much everything else, really), so I think there were some compilers which didn't allow leading whitespace, but I don't think I've ever actually run into one in the wild, though.

Reply:
Sorry, as I said in my edit, you are right.
However at work, post 2000, I had an unforgiving compiler on one of the historic platforms for our multi-platform software.
I think it was probably the compiler for the DEC Alpha on their Tru64 Unix system V4.0F.
(Incidentally, at the time, the Alpha was a wonderful processor, blowing away anything from Intel/AMD.)
However it just could have been the SGI compiler on Irix.

Reply:
Sorry, I didn't see the edit, just read the e-mail.  That makes sense; I've only been using C since the late 90s, and then I was really only using GCC (IIRC I actually started with egcs, which should give you a good idea of the timeframe), so while I heard people say the # had to be at the beginning of the line it was never a problem for me personally.

Anyways, back to the masking stuff.  I was looking into this a bit further, and it looks like it's actually the compilers optimizing anything > 63.  See <https://godbolt.org/z/2thNFh>.  MSVC is the only compiler that doesn't optimize anything greater than 63 to an xor of itself (i.e., all zeros).  And what MSVC does generate for values > 63 is insane; instead of just using the value as an immediate mode argument to vpsrlq they move it into a vector register then use that.

Reply:
Thanks for pointing me to Compiler Explorer: I didn't know it.
I tried the following code where it cannot possibly know the shift count and I can see no > 63 test.

```
#include <immintrin.h>
#include <stdlib.h>

__m512i foo(__m512i bar) {
    unsigned int j = rand();
    return _mm512_srli_epi64(bar, j);
}

```

Reply:
Enjoy Compiler Explorer; it's a fantastic tool.  FWIW, you can use SIMDe on it: http://simde.netlify.com/godbolt/simple is the basic one, or http://simde.netlify.com/godbolt/demo to see a bunch of compilers. 

Good point about making sure the compiler can't know the value; I'm used to the intrinsic requiring that argument be a compile-time constant, so I didn't even think of it.  If I do

```c
#include <immintrin.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdio.h>

__m512i foo(__m512i bar, unsigned int j) {
    return _mm512_srli_epi64(bar, j);
}

int main(int argc, char** argv) {
  __m512i v = _mm512_set1_epi64(~UINT64_C(0));
  __m512i r = _mm512_srli_epi64(v, atoi(argv[1]));
  printf("0x%llx\n", ((uint64_t*) &r)[0]);

  return 0;
}
```

It uses all 64 bits, which makes sense; it matches the behavior of `_mm512_srl_epi64`.  I'm now convinced that the intrinsics guide is wrong.